### PR TITLE
Feature/Enemies&Bosses statemachine

### PR DESCRIPTION
**Pull request by bas-boop**

### CODE REVIEW

File: Assets/Scripts/Framework/Statemachine/BaseState.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+using UnityEngine;
+
+public abstract class BaseState : MonoBehaviour
+{
+    protected StateMachine Parent;
+    public bool IsValidToSwitch { get; protected set; }
+
+    public void SetParent(StateMachine targetMachine) => Parent = targetMachine;
+
+    public virtual void EnterState(StateMachine stateMachine) => IsValidToSwitch = false;
```
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/BaseState.cs]: Sure you want to make EnterState virtual? Seems like it's a rather important method for it to be optional.

File: Assets/Scripts/Framework/Statemachine/StateMachine.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public abstract class StateMachine : MonoBehaviour
```
Comment (Tatum13) [inline Assets/Scripts/Framework/Statemachine/StateMachine.cs]:
```suggestion
public abstract class FiniteStateMachine : MonoBehaviour
```

File: Assets/Scripts/Framework/Statemachine/SmallEnmiesBaseState.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+public abstract class SmallEnmiesBaseState : BaseState
```
Comment (Tatum13) [inline Assets/Scripts/Framework/Statemachine/SmallEnmiesBaseState.cs]:
```suggestion
public abstract class SmallEnemiesBaseState : BaseState
```

File: Assets/Scripts/Framework/Statemachine/SmallEnmiesBaseState.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+{
+    #region BaseState to SmallEnmiesState
```
Comment (Tatum13) [inline Assets/Scripts/Framework/Statemachine/SmallEnmiesBaseState.cs]:
```suggestion
    #region BaseState to SmallEnemiesState
```

File: Assets/Scripts/Framework/Statemachine/BigEnmiesStateMachine.cs [hunk 5]
```diff
@@ -0,0 +1,4 @@
+public class BigEnmiesStateMachine : StateMachine
```
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/BigEnmiesStateMachine.cs]: May wanna fix that class name there big man

File: Assets/Scripts/Framework/Statemachine/BossBaseState.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public abstract class BossBaseState : BaseState
```
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/BossBaseState.cs]: I like that you have abstract classes inherit from abstract classes. Very nice.

File: Assets/Scripts/Framework/Statemachine/BossBaseState.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+{
+    #region BaseState to BigEnmiesState
```
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/BossBaseState.cs]: Again, may wanna fix that spelling there xd.

Comment (Tatum13) [inline Assets/Scripts/Framework/Statemachine/BossBaseState.cs]:
```suggestion
    #region BaseState to BigEnemiesState
```

File: Assets/Scripts/Framework/Statemachine/BigEnmiesStateMachine.cs (hunk 5 above)
Comment (Tatum13) [inline Assets/Scripts/Framework/Statemachine/BigEnmiesStateMachine.cs]:
```suggestion
public class BigEnemiesStateMachine : StateMachine
```

File: Assets/Scripts/Framework/Statemachine/BossStateMachine.cs [hunk 8]
```diff
@@ -0,0 +1,18 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class BossStateMachine : StateMachine
+{
+    // Start is called before the first frame update
```
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/BossStateMachine.cs]: Is it huh? Interesting! (Fr tho why is this here if it's just empty)

File: Assets/Scripts/Framework/Statemachine/BigEnmiesBaseState.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
+public abstract class BigEnmiesBaseState : BaseState
+{
+    #region BaseState to BigEnmiesState
```
Comment (Tatum13) [inline Assets/Scripts/Framework/Statemachine/BigEnmiesBaseState.cs]:
```suggestion
    #region BaseState to BigEnemiesState
```

File: Assets/Scripts/Framework/Statemachine/BigEnmiesBaseState.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
+public abstract class BigEnmiesBaseState : BaseState
```
Comment (Tatum13) [inline Assets/Scripts/Framework/Statemachine/BigEnmiesBaseState.cs]:
```suggestion
public abstract class BigEnemiesBaseState : BaseState
```

File: Assets/Scripts/Framework/Statemachine/BaseState.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+using UnityEngine;
+
+public abstract class BaseState : MonoBehaviour
```
Comment (Tatum13) [inline Assets/Scripts/Framework/Statemachine/BaseState.cs]:
```suggestion
public abstract class FiniteBaseState : MonoBehaviour
```

**Review by Tatum13 — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Framework/Statemachine/SmallEnmiesBaseState.cs (hunk 3 above)
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/SmallEnmiesBaseState.cs]: It's "Enemies", hope that helps :)

**Comment by Tatum13** [comment]
Verander de branch name naar : feature/enemies-and-bosses-statemachine

File: Assets/Scripts/Framework/Statemachine/SmallEnmiesBaseState.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+    public override void EnterState(StateMachine entity)
+    {
+        base.EnterState(Parent);
+        EnterState((SmallEnemiesStateMachine) Parent);
```
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/SmallEnmiesBaseState.cs]: I do wonder though, why is a Parent something that is constantly being used as a parameter for methods such as "EnterState"?

File: Assets/Scripts/Framework/Statemachine/StateMachine.cs [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        currentState.FixedUpdateState(this);
+
+        if (!currentState.IsValidToSwitch || _switchStateQueue.Count <= 0) return;
+        
+        SwitchState(_switchStateQueue[0]);
+        _switchStateQueue.Remove(_switchStateQueue[0]);
+    }
+
+    private void SetStatesParent()
```
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/StateMachine.cs]: Instead of having to go over a list of states to make yourself their parent, why not the other way around? Allow states to make this statemachine their own parent.

I refer you to the Single Responsibility Principle ;)

File: Assets/Scripts/Framework/Statemachine/StateMachine.cs [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        currentState.EnterState(this);
+    }
+
+    /// <summary>
+    /// Switch the current targetState to a different one.
+    /// Is it valid to switch targetState?
+    /// </summary>
+    /// <param name="targetState">Give target state to switch into.</param>
+    /// <param name="isCalledFromExitState">If the SwitchState is called form a ExitState function, this needs to be true!</param>
+    public void SwitchState(BaseState targetState, bool isCalledFromExitState = false)
```
Comment (BNS-MarkUlrich) [inline Assets/Scripts/Framework/Statemachine/StateMachine.cs]: Why is this the only method so far with a summary? Why specifically this one?

**Review by BNS-MarkUlrich — CHANGES_REQUESTED** [review]
Looks pretty good for a first look-over, but no cigar yet

File: Assets/Scripts/Framework/Statemachine/BaseState.cs (hunk 1 above)
Comment (Tatum13) [reply Assets/Scripts/Framework/Statemachine/BaseState.cs]: make it abstract? @BNS-MarkUlrich

File: Assets/Scripts/Framework/Statemachine/StateMachine.cs (hunk 14 above)
Comment (bas-boop) [reply Assets/Scripts/Framework/Statemachine/StateMachine.cs]: Omdat de parameter misschien confussing kan zijn. Had tijdens het gebruik wat error's dus moest het op een manier fixen.

Comment (bas-boop) [reply Assets/Scripts/Framework/Statemachine/StateMachine.cs]: Wil je het weg dan?

Comment (BNS-MarkUlrich) [reply Assets/Scripts/Framework/Statemachine/StateMachine.cs]: No it's fine, I was just wondering why this one and not other methods!

File: Assets/Scripts/Framework/Statemachine/BaseState.cs (hunk 1 above)
Comment (bas-boop) [reply Assets/Scripts/Framework/Statemachine/BaseState.cs]: Dan mis je wat er gebeurd in de body.

Comment (BNS-MarkUlrich) [reply Assets/Scripts/Framework/Statemachine/BaseState.cs]: Yes. Seeing as all the other methods like it are, I would say it makes sense as EnterState and ExitState are probably the most important methods of BaseState and any inheritors absolutely cannot forget to implement it or by accident create something to bypass EnterState, wouldn't want that!

Comment (BNS-MarkUlrich) [reply Assets/Scripts/Framework/Statemachine/BaseState.cs]: > Dan mis je wat er gebeurd in de body.

I'm not sure you should want a check like that though. If you do that only really means you've created an edge case that should be avoidable

**Review by BNS-MarkUlrich — APPROVED** [review]

**Review by Tatum13 — APPROVED** [review]
